### PR TITLE
Tutorial example: Allocator template

### PR DESCRIPTION
**Pull request by jacquelinekay**
This is *not* a requirement to get shipped in Alpha 2, especially since running the code itself is extremely unimpressive. I just want to get review for this code since I've put it in a tutorial:

https://github.com/ros2/ros2/wiki/Allocator-Template-Tutorial

### CODE REVIEW

File: rclcpp_examples/src/topics/allocator_example.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include <rclcpp/rclcpp.hpp>
+#include <std_msgs/msg/u_int32.hpp>
```
Comment (dirk-thomas) [inline rclcpp_examples/src/topics/allocator_example.cpp]: cpplint

File: rclcpp_examples/src/topics/allocator_example.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  // message on the execution path, it will use the custom deallocate.
+  auto msg = std::allocate_shared<std_msgs::msg::UInt32>(*alloc.get());
+
+  rclcpp::utilities::sleep_for(std::chrono::milliseconds(1));
+  is_running = true;
+
+  uint32_t i = 0;
+  while (rclcpp::ok()) {
+    msg->data = i;
+    i++;
```
Comment (dirk-thomas) [inline rclcpp_examples/src/topics/allocator_example.cpp]: ++i

File: rclcpp_examples/src/topics/allocator_example.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    publisher->publish(msg);
+    rclcpp::utilities::sleep_for(std::chrono::milliseconds(1));
+    executor.spin_some();
+  }
+  is_running = false;
+
+  uint32_t final_global_allocs = global_runtime_allocs;
+  uint32_t final_global_deallocs = global_runtime_deallocs;
+  std::cout << "Global new was called " << final_global_allocs
```
Comment (dirk-thomas) [inline rclcpp_examples/src/topics/allocator_example.cpp]: printf

File: rclcpp_examples/src/topics/allocator_example.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            << " times during spin" << std::endl;
+  std::cout << "Global delete was called " << final_global_deallocs
+            << " times during spin" << std::endl;
+
+  std::cout << "Allocator new was called " << num_allocs
+            << " times during spin" << std::endl;
+  std::cout << "Allocator delete was called " << num_deallocs
+            << " times during spin" << std::endl;
+  rclcpp::utilities::sleep_for(std::chrono::milliseconds(1));
```
Comment (dirk-thomas) [inline rclcpp_examples/src/topics/allocator_example.cpp]: Why sleep here?

File: rclcpp_examples/src/topics/allocator_example.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  // Constructs the intra-process manager with a custom allocator.
+  context->get_sub_context<rclcpp::intra_process_manager::IntraProcessManager>(ipm_state);
+  auto node = rclcpp::Node::make_shared("allocator_example", true);
+  */
+
+  uint32_t counter = 0;
+  auto callback = [&counter](std_msgs::msg::UInt32::SharedPtr msg) -> void
+    {
+      (void)msg;
+      counter++;
```
Comment (dirk-thomas) [inline rclcpp_examples/src/topics/allocator_example.cpp]: ++counter

**Comment by dirk-thomas** [comment]
Beside the comments: lgtm

File: rclcpp_examples/src/topics/allocator_example.cpp (hunk 4 above)
Comment (jacquelinekay) [inline rclcpp_examples/src/topics/allocator_example.cpp]: Sometimes I see the program exit before these statements print to screen. Maybe when I change them to printfs this problem will be fixed.

**Comment by wjwwood** [comment]
lgtm.  A squash of the commits would be nice.

**Pull request merged**